src/main/host/process.rs

```diff
@@ -0,0 +1,215 @@
+}
+
+impl Process {
+    /// For now, this should only be called via Worker, to borrow the current
+    /// Process. This will ensure there is only one reference to a given Process
+    /// in Rust.
+    ///
+    /// SAFETY: `p` must point to a valid c::Process, to which this Process will
+    /// have exclusive access over its lifetime. `p` must outlive the returned object.
+    pub unsafe fn borrow_from_c(p: *mut cshadow::Process) -> Self {
+        assert!(!p.is_null());
+        Process { cprocess: p }
+    }
+
+    /// Read data from `src` in the current Process into a potentially-uninitialized `dst`.
+    pub fn read_ptr_into_slice<T: Pod>(&self, dst: &mut [T], src: PluginPtr) -> nix::Result<()> {
+        let status = unsafe {
+            cshadow::process_readPtr(
+                self.cprocess,
+                dst.as_mut_ptr() as *mut std::os::raw::c_void,
+                src.into(),
+                (dst.len() * std::mem::size_of::<T>()) as u64,
+            )
+        };
+        if status == 0 {
+            Ok(())
+        } else {
+            Err(nix::Error::from_errno(nix::errno::from_i32(status as i32)))
+        }
+    }
+
+    /// Read data from `src` in the current Process into `dst`.
+    pub fn read_ptr_into_val<'a, T: Pod>(&self, dst: &mut T, src: PluginPtr) -> nix::Result<()> {
+        let mut slice = std::slice::from_mut(dst);
+        self.read_ptr_into_slice(&mut slice, src)?;
+        Ok(())
+    }
+
+    /// Interpret `src` as a pointer to `T` in the current Process in `dst` and return its value.
+    pub fn read_ptr_as_val<T: Pod>(&self, src: PluginPtr) -> nix::Result<T> {
+        // SAFETY: Since `T` is Pod, any bit pattern is a valid `T`.
+        let mut val: T = unsafe { MaybeUninit::uninit().assume_init() };
+        self.read_ptr_into_val(&mut val, src)?;
+        Ok(val)
+    }
+
+    /// Write `src` into `dst` in the current Process.
+    pub fn write_ptr_from_slice<T: Copy>(&mut self, dst: PluginPtr, src: &[T]) -> nix::Result<()> {
+        let status = unsafe {
+            cshadow::process_writePtr(
+                self.cprocess,
+                dst.into(),
+                src.as_ptr() as *const std::os::raw::c_void,
+                (src.len() * std::mem::size_of::<T>()) as u64,
+            )
+        };
+        if status == 0 {
+            Ok(())
+        } else {
+            Err(nix::Error::from_errno(nix::errno::from_i32(status as i32)))
+        }
+    }
+
+    /// Write `src` into `dst` in the current Process.
+    pub fn write_ptr_from_val<T: Copy>(&mut self, dst: PluginPtr, src: &T) -> nix::Result<()> {
+        let src = std::slice::from_ref(src);
+        self.write_ptr_from_slice(dst, src)
+    }
+
+    // For simplicity, never returns NULL, even for s == 0.
+    fn get_readable_ptr(
+        &self,
+        p: PluginPtr,
+        s: usize,
+    ) -> nix::Result<*const ::std::os::raw::c_void> {
+        if s == 0 {
+            return Ok(std::ptr::null());
+        }
```

Comment:
This comment is wrong, and this will result in UB when `s == 0` since `get_slice()` will call `std::slice::from_raw_parts()` with a NULL pointer.